### PR TITLE
[ARROW] Vectorized Parquet Reads - Assertions + Correctness Checks for Deletion Files

### PR DESCRIPTION
**Pull request by mayursrivastava**
Changes to the ArrowReader:
1. Add assertions for reader support
2. Don't process delete files since they are not supported in the reader
3. Fix an error message
4. Fix the code style

### CODE REVIEW

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java [hunk 1]
```diff
@@ -84,11 +92,26 @@
...
+          TypeID.INTEGER,
+          TypeID.LONG,
+          TypeID.FLOAT,
+          TypeID.DOUBLE,
+          TypeID.STRING,
+          TypeID.TIMESTAMP,
+          TypeID.BINARY,
+          TypeID.DATE,
+          TypeID.UUID,
+          TypeID.TIME
```
Comment (rdblue) [inline arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java:113]: Indentation is off throughout this PR. It should be 2 spaces for an indent and 2 indents for a continuation. Here, you should use 4 spaces = 2 indents.

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java [hunk 2]
```diff
@@ -136,6 +159,14 @@ public ArrowReader(TableScan scan, int batchSize, boolean reuseContainers) {
    * previous {@link ColumnarBatch} may be reused for the next {@link ColumnarBatch}.
    * This implies that the caller should either use the {@link ColumnarBatch} or deep copy the
    * {@link ColumnarBatch} before getting the next {@link ColumnarBatch}.
+   *
+   * <p>This method works for only when the following conditions are true:
```
Comment (rdblue) [inline arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java:163]: Nit: `<p>` normally goes on the line between paragraphs.

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java [hunk 3]
```diff
@@ -197,26 +227,41 @@ public void close() throws IOException {
...
-          .flatMap(Collection::stream)
-          .collect(Collectors.toList());
+    VectorizedCombinedScanIterator(CloseableIterable<CombinedScanTask> tasks,
+                                   Schema expectedSchema,
+                                   String nameMapping,
+                                   FileIO io,
+                                   EncryptionManager encryptionManager,
+                                   boolean caseSensitive,
+                                   int batchSize,
+                                   boolean reuseContainers) {
```
Comment (rdblue) [inline arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java:237]: This format is closer to what we would normally use for method arguments, but a more important rule is not making changes larger than necessary by changing whitespace. Could you remove the reformatting here and for other argument lists?

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java [hunk 4]
```diff
@@ -197,26 +227,41 @@ public void close() throws IOException {
...
+              .map(c -> c.type().typeId())
+              .allMatch(SUPPORTED_TYPES::contains);
+      if (!atLeastOneColumn || !hasNoDeleteFiles || !hasSupportedTypes) {
+        throw new UnsupportedOperationException(
+                "ArrowReader is supported for the query schema with at least one column," +
+                        " with no delete files and for supported data types" +
+                        ", but found that atLeastOneColumn=" + atLeastOneColumn +
+                        ", hasNoDeleteFiles=" + hasNoDeleteFiles +
+                        ", hasSupportedTypes=" + hasSupportedTypes +
+                        ", supported types=" + SUPPORTED_TYPES +
```
Comment (rdblue) [inline arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java:256]: This error message isn't very helpful because it checks several things and then mixes them together so the person reading the error has to figure out what is already known here: whether the failure was because of delete files, supported types, or expected columns.

This should be reformatted into 3 different checks with specific error messages, like "Cannot read files that require applying delete files: <split>", "Cannot read without at least one projected column", and "Cannot read unsupported column types: <unsupported-types>". The last one should produce a list of the types that are used but not supported, since that is already known.

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java [hunk 5]
```diff
@@ -197,26 +227,41 @@ public void close() throws IOException {
...
+                        ", expected columns=" + expectedSchema.columns()
+        );
+      }
+
       Map<String, ByteBuffer> keyMetadata = Maps.newHashMap();
       fileTasks.stream()
-          .flatMap(fileScanTask -> Stream.concat(Stream.of(fileScanTask.file()), fileScanTask.deletes().stream()))
+          .map(FileScanTask::file)
           .forEach(file -> keyMetadata.put(file.path().toString(), file.keyMetadata()));
-
```
Comment (rdblue) [inline arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java:219]: Nit: unnecessary whitespace change.

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java [hunk 6]
```diff
@@ -252,17 +297,11 @@ public boolean hasNext() {
...
-              e);
-        } else {
-          throw new RuntimeException(
-              "An error occurred while iterating through the file scan tasks or closing the iterator," +
-                  " see the stacktrace for further information. The file scan tasks are: " + fileTasks, e);
+          LOG.error("Error reading file: {}", getInputFile(currentTask).location(), e);
         }
+        // Wrap and throw a Runtime exception because Iterator::hasNext()
+        // cannot throw a checked exception.
+        throw new RuntimeException(e);
```
Comment (rdblue) [inline arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java:304]: If the exception is already a `RuntimeException` then this should not wrap it in another because that may change the exception type. You can use `ExceptionUtil.castAndThrow(e, RuntimeException.class);` for this.

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java [hunk 7]
```diff
@@ -323,10 +362,9 @@ private InputFile getInputFile(FileScanTask task) {
      * @param fileSchema             Schema of the data file.
      * @param setArrowValidityVector Indicates whether to set the validity vector in Arrow vectors.
      */
-    private static ArrowBatchReader buildReader(
-        Schema expectedSchema,
-        MessageType fileSchema,
-        boolean setArrowValidityVector) {
+    private static ArrowBatchReader buildReader(Schema expectedSchema,
+                                                MessageType fileSchema,
+                                                boolean setArrowValidityVector) {
```
Comment (rdblue) [inline arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java:367]: This looks like another change that doesn't need to be done.

**Comment by rdblue** [comment]
Thanks, @mayursrivastava! I found a few things to fix, but it should be mostly minor updates.

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java (hunk 6 above)
Comment (mayursrivastava) [reply arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java]: Agreed. ExceptionUtil::castAndThrow was not public and to use it here, I've made it public. I hope that is ok.

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java (hunk 1 above)
Comment (mayursrivastava) [reply arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java]: fixed

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java (hunk 2 above)
Comment (mayursrivastava) [reply arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java]: fixed

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java (hunk 3 above)
Comment (mayursrivastava) [reply arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java]: done

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java (hunk 4 above)
Comment (mayursrivastava) [reply arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java]: Fixed

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java (hunk 5 above)
Comment (mayursrivastava) [reply arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java:219]: fixed

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java (hunk 7 above)
Comment (mayursrivastava) [reply arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java]: fixed

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java [hunk 8]
```diff
@@ -252,17 +297,10 @@ public boolean hasNext() {
         }
       } catch (IOException | RuntimeException e) {
         if (currentTask != null && !currentTask.isDataTask()) {
-          throw new RuntimeException(
```
Comment (nastra) [inline arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java:255]: while reviewing this part of the PR, I noticed that `hasNext` isn't idempotent, meaning that if you call `hasNext` multiple times, it will actually advance the pointer and IMO this kind of logic should only happen inside `next` and not `hasNext`.

I also wonder why we need to catch exceptions in a `hasNext` method, so maybe this is something that can be improved in a follow-up?

**Review by nastra — APPROVED** [review]
the proposed changes LGTM, but I think `hasNext` & `next` logic should be improved in a follow-up

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java (hunk 8 above)
Comment (mayursrivastava) [reply arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java:255]: Good catch. I'll make hasNext() idempotent in a follow-up cr. 

On the point of exceptions, I think exceptions should not cause iteration to silently stop and it is better to raise them. There are other instances in Iceberg where similar exception handling is done, for example: https://github.com/apache/iceberg/blob/90225d6c9413016d611e2ce5eff37db1bc1b4fc5/api/src/main/java/org/apache/iceberg/io/CloseableIterable.java#L156. Let me know if I understand your comment correctly. I'm happy to discuss this further and improve the code accordingly.

Comment (nastra) [reply arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java:255]: I was just curious whether this kind of exception handling was necessary here or not. I guess once `hasNext` is idempotent, the exception handling (if necessary) can be moved to `next`

**Comment by kbendick** [comment]
Thank you for your contribution @mayursrivastava! One small nit: can you update the PR title to be a bit more descriptive?

I'm particularly interested in the vectorized parquet read paths, but the PR title alone doesn't convey much info to me. Maybe `[ARROW] Vectorized Parquet Reads - Assertions + Correctness Checks for Deletion Files`? Will leave that up to you as it's your PR, but it would be great if you could update the PR title :slightly

**Comment by kbendick** [comment]
So SORRY! I fat fingered the close PR button when trying to use an emoji 🙃 

Thank you for your contribution @mayursrivastava! One small nit: can you update the PR title to be a bit more descriptive?

I'm particularly interested in the vectorized parquet read paths, but the PR title alone doesn't convey much info to me. Maybe [ARROW] Vectorized Parquet Reads - Assertions + Correctness Checks for Deletion Files? Will leave that up to you as it's your PR, but it would be great if you could update the PR title 🙂

**Comment by mayursrivastava** [comment]
@kbendick Thanks! I changed the pr title

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java (hunk 8 above)
Comment (mayursrivastava) [reply arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java:255]: The change will be the following:
`--- a/arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java
+++ b/arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java
@@ -25,6 +25,7 @@ import java.util.Collection;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+import java.util.NoSuchElementException;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import java.util.stream.StreamSupport;
@@ -172,7 +173,6 @@ public class ArrowReader extends CloseableGroup {
     private final int batchSize;
     private final boolean reuseContainers;
     private CloseableIterator<ColumnarBatch> currentIterator;
-    private ColumnarBatch current;
     private FileScanTask currentTask;
 
     /**
@@ -239,7 +239,6 @@ public class ArrowReader extends CloseableGroup {
       try {
         while (true) {
           if (currentIterator.hasNext()) {
-            this.current = currentIterator.next();
             return true;
           } else if (fileItr.hasNext()) {
             this.currentIterator.close();
@@ -268,7 +267,11 @@ public class ArrowReader extends CloseableGroup {
 
     @Override
     public ColumnarBatch next() {
-      return current;
+      if (currentIterator.hasNext()) {
+        return currentIterator.next();
+      } else {
+        throw new NoSuchElementException();
+      }
     }`
The exception handling will not change. This is similar to he CloseableIterable.hasNext().

**Comment by mayursrivastava** [comment]
How can I trigger the Java CI that got canceled?

**Comment by szehon-ho** [comment]
Not sure, but hacky way, can amend last commit with no change and force push

**Comment by mayursrivastava** [comment]
Hi @rdblue, is this pr good to go? One of the tests timed out. Is it possible for admins to rerun it or should I make an amend to the commit and force push (as recommended earlier on this thread)?

**Comment by mayursrivastava** [comment]
@rdblue is this pr ready to be merged?

File: core/src/main/java/org/apache/iceberg/util/ExceptionUtil.java [hunk 9]
```diff
@@ -29,7 +29,7 @@ private ExceptionUtil() {
   }
 
   @SuppressWarnings("unchecked")
-  static <E extends Exception> void castAndThrow(
+  public static <E extends Exception> void castAndThrow(
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/util/ExceptionUtil.java:32]: I think this was recently changed in a different PR, so you may need to rebase.

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java [hunk 10]
```diff
@@ -206,15 +238,28 @@ public void close() throws IOException {
...
+      List<FileScanTask> fileTasks = StreamSupport.stream(tasks.spliterator(), false)
           .map(CombinedScanTask::files)
           .flatMap(Collection::stream)
           .collect(Collectors.toList());
       this.fileItr = fileTasks.iterator();
 
+      if (fileTasks.stream().anyMatch(TableScanUtil::hasDeletes)) {
+        throw new UnsupportedOperationException("Cannot read files that require applying delete files");
+      }
+      if (expectedSchema.columns().isEmpty()) {
```
Comment (rdblue) [inline arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java:250]: Style nit: Iceberg adds an empty newline after control flow statements like `if`/`else`/`for`/`while`.

**Review by rdblue — APPROVED** [review]
Overall, this looks good. I think it needs to be rebased and then we can merge.

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java (hunk 10 above)
Comment (mayursrivastava) [reply arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java:251]: fixed

**Comment by mayursrivastava** [comment]
@rdblue rebased

**Comment by mayur-srivastava** [comment]
@rdblue could we get this merged? Let me know if anything else is needed. Thanks.

**Comment by rdblue** [comment]
Re-running CI checks.

**Comment by rdblue** [comment]
Thanks, @mayursrivastava!
